File: java/06_003.java
import java.util.Arrays;

public class QuickSort {
  /**
   * Sort array using quick sort algorithm
   */
  public static void quickSort(int[] arr, int low, int high) {
     if (low < high) {
        // Get partition index
        int pi = partition(arr, low, high);
        
        // Sort elements before and after partition
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
     }
  }
  
  /**
   * Partition array around pivot element
   */
  private static int partition(int[] arr, int low, 
        int high) {
     // Select rightmost element as pivot
     int pivot = arr[high];
     int i = (low - 1);
     
     // Move elements smaller than pivot to left
     for (int j = low; j < high; j++) {
        if (arr[j] <= pivot) {
           i++;
           
           // Swap elements
           int temp = arr[i];
           arr[i] = arr[j];
           arr[j] = temp;
        }
     }
     
     // Place pivot in correct position
     int temp = arr[i + 1];
     arr[i + 1] = arr[high];
     arr[high] = temp;
     
     return i + 1;
  }
  
  public static void main(String[] args) {
     int[] numbers = {23, 16, 6, 59, 3, 11, 37};
     System.out.println("Original array: " + 
        Arrays.toString(numbers));
     quickSort(numbers, 0, numbers.length - 1);
     System.out.println("Sorted array: " + 
        Arrays.toString(numbers));
  }
}

File: cpp/06_003.cpp
#include <iostream>
#include <vector>

void quickSort(std::vector<int>& arr, int low, int high) {
    if (low < high) {
        // Partition function
        auto partition = [&arr](int low, int high) {
            int pivot = arr[high];
            int i = low - 1;
            
            for (int j = low; j < high; j++) {
                if (arr[j] <= pivot) {
                    i++;
                    std::swap(arr[i], arr[j]);
                }
            }
            
            std::swap(arr[i + 1], arr[high]);
            return i + 1;
        };
        
        int pi = partition(low, high);
        
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

int main() {
    std::vector<int> numbers = {23, 16, 6, 59, 3, 11, 37};
    
    std::cout << "Original array: ";
    for (int num : numbers) std::cout << num << " ";
    std::cout << std::endl;
    
    quickSort(numbers, 0, numbers.size() - 1);
    
    std::cout << "Sorted array: ";
    for (int num : numbers) std::cout << num << " ";
    std::cout << std::endl;
    
    return 0;
}